### PR TITLE
Add new LABEL annotation type

### LINKED ISSUE #26: Text annotations
Good stuff, you sure saved me a couple hours! Any updates on Text annotations?

Additionally, wouldn't it be neat to attach annotations to datasets? That way, annotations can be hidden when its related dataset is toggled.

Has this project been abandoned, or are there any plans for these updates or similar?

Thanks :)

### PR DESCRIPTION
**Pull request by stockiNail**
Fixes   #26

Creates new `LabelAnnotation` element to enable to render a text on the chart at a coordinate, calculated by the chart data values.

Sets the following options:

| Name | Description |
| ---- | ---- |
| `align` | Specifies the text horizontal position of the label with respect to the selected point.
| `adjustScaleRange` | Should the scale range be adjusted if this annotation is out of range.
| `backgroundColor` | Fill color.
| `borderColor` | Stroke color.
| `borderWidth` | Stroke width.
| `borderDash` | Length and spacing of dashes.
| `borderDashOffset` | Offset for line dashes.
| `borderRadius` | Radius of the corners of the annotation.
| `color` | Text color.
| `content` | The content to show in the text annotation.
| `display` | Whether or not this annotation is visible.
| `drawTime` | Determines where in the chart lifecycle the drawing occurs.
| `font` | Text font.
| `position` | Specifies the text vertical position of the label with respect to the selected point.
| `textAlign` | Text alignment of label content when there's more than one line.
| `xAdjust` | Adjustment along x-axis (left-right) of label relative to computed position.
| `xPadding` | Padding of label to add left/right.
| `xScaleID` | ID of the X scale to bind onto.
| `xValue` | X coordinate of the point in units along the x axis.
| `yAdjust` | Adjustment along y-axis (top-bottom) of label relative to computed position.
| `yPadding` | Padding of label to add top/bottom.
| `yScaleID` | ID of the Y scale to bind onto.
| `yValue` | Y coordinate of the point in units along the y axis.

Being new annotation type, you have the events on that as well.

- [X] test cases
- [x] documentation

### CODE REVIEW

**Comment by stockiNail** [comment]
`rotation` option and `content` as image will be implemented later, need more time

File: docs/.vuepress/config.js [hunk 1]
```diff
@@ -1,78 +1,80 @@
...
-          children: [
-            'charts/bar',
-            'charts/line',
-          ],
-        },
-      ]
-    }
-  }
-};
+module.exports = {
```
Comment (kurkle) [inline docs/.vuepress/config.js:1]: Not quite sure what the whitespace change is in this file?

File: src/types/text.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+const widthCache = new Map();
+function measureLabel(ctx, label) {
+  const content = label.content;
+  const lines = isArray(content) ? content : [content];
+  const count = lines.length;
+  const font = toFont(label.font);
+  const lh = font.lineHeight;
+  const xPadding = label.xPadding;
+  const yPadding = label.yPadding;
+  const borderWidth = label.borderWidth;
```
Comment (kurkle) [inline src/types/text.js:124]: These are only used once, so no need to cache in variables.

**Review by kurkle — COMMENTED** [review]
Also the comments from box annotation label are duplicated here.

I'd really like to have the line and box annotation use TextAnnotation internally for label, it would remove the code duplication and simplify maintenance.

File: docs/.vuepress/config.js (hunk 1 above)
Comment (stockiNail) [reply docs/.vuepress/config.js:1]: no.. :( I'm gonna remove it

File: src/types/text.js (hunk 2 above)
Comment (stockiNail) [reply src/types/text.js]: done!

**Comment by stockiNail** [comment]
> I'd really like to have the line and box annotation use TextAnnotation internally for label, it would remove the code duplication and simplify maintenance.

It makes sense and I thought about that before starting implementation.

Nevertheless text, box and line annotations have got differences to evaluate:

1. BOX is getting the values by xMin, xMax, yMin and yMax (rect). TEXT by xValue and yValue (point). Furthermore label is node of BOX annotation, TEXT not.
2. LINE is getting the values by several options but not xValue and yValue. Furthermore label is node of LINE annotation, TEXT not. And LABEL of LINE is strictly related to the line (similar to TEXT but not to BOX).

Let me propose you to go ahead with this implementation and in the meanwhile I'll prepare new PR with the inheritance among the classes (class diagram).

Make sense?

File: docs/.vuepress/config.js (hunk 1 above)
Comment (stockiNail) [reply docs/.vuepress/config.js:1]: Honestly, I was thinking it was related to last row (empty) but it's not... I don't why it's still there.

Comment (kurkle) [reply docs/.vuepress/config.js:1]: line endigs?

Comment (stockiNail) [reply docs/.vuepress/config.js:1]: Tomorrow I’ll check

Comment (stockiNail) [reply docs/.vuepress/config.js]: Fixed

**Comment by kurkle** [comment]
So which way was the line ending changed? Editorconfig says it should be `lf` and maybe your editor just did that?

**Comment by stockiNail** [comment]
> So which way was the line ending changed? Editorconfig says it should be `lf` and maybe your editor just did that?

I’m using Eclipse as IDE and this is the reason.
For some file types when I press Enter it adds the line separator

**Comment by stockiNail** [comment]
> I'd really like to have the line and box annotation use TextAnnotation internally for label, it would remove the code duplication and simplify maintenance.

@kurkle added discussion #480 in order to maintain in a separated place the discussion about the re-factoring with inheritance among annotation elements. Let me know what you think.

**Comment by stockiNail** [comment]
@kurkle if you agree, I'd like to rename the class from TextAnnotation to LabelAnnotation because, going to the discussion #480, this annotation type will manage also images and not only texts.
Therefore I'd ask you to wait for approval.

**Comment by stockiNail** [comment]
@kurkle FYI I have started some coding to address the discussion #480, branch https://github.com/pepstock-org/chartjs-plugin-annotation/tree/inheritance

**Comment by kurkle** [comment]
> @kurkle FYI I have started some coding to address the discussion #480, branch https://github.com/pepstock-org/chartjs-plugin-annotation/tree/inheritance
> 
> The branch was started from master, merging the 3 PRs, currently open in chartjs/chartjs-plugin-annotation project

I'm not sure if we should use inheritance in elements (or if it works properly with registration).

I think a generic "label" element (or class) is what we really need, and it should be moved to core eventually.

Lets see what you come up with first and go from there :)

**Comment by stockiNail** [comment]
> (or if it works properly with registration).

Sorry, I don't know exactly what you meant with 'registration'.

**Comment by kurkle** [comment]
> > (or if it works properly with registration).
> 
> Sorry, I don't know exactly what you meant with 'registration'.

That is what I had in mind: (but is probably not a problem)

https://github.com/chartjs/Chart.js/blob/2988a6c6dcb582e29ab96b3947cf2565a97da261/src/core/core.typedRegistry.js#L17

**Comment by stockiNail** [comment]
@kurkle I spent more time than I expected to think how to implement.

Finally I decided to use class inheritance because, in my poor opinion, sounds better.

Therefore I write the code where BoxAnnotation extends LabelAnnotation which extends PointAnnotation.

A Label class is created (labelItem.js) to manage the labelling work.

Here is committed: https://github.com/pepstock-org/chartjs-plugin-annotation/tree/inheritance

What is missing is the changes on LineAnnotation in order to use common code also for that.

I'm going on but if you have time to have a look and stop me if you don't like.

**Comment by kurkle** [comment]
| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| `dist/chartjs-plugin-annotation.esm.js` | 8.45 kB | +868 B (+11%) | ⚠️ |
| `dist/chartjs-plugin-annotation.js` | 8.58 kB | +876 B (+11%) | ⚠️ |
| `dist/chartjs-plugin-annotation.min.js` | 6.03 kB | +640 B (+12%) | ⚠️ |

**Comment by stockiNail** [comment]
Closed in favor of #502

**Pull request closed without merging**